tests/integration/targets/terraform/tasks/complex_variables.yml

```
@@ -0,0 +1,56 @@
+---
+# Copyright (c) Ansible Project
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+- name: Create terraform project directory (complex variables)
+  file:
+    path: "{{ terraform_project_dir }}/complex_vars"
+    state: directory
+    mode: 0755
+
+- name: copy terraform files to work space
+  ansible.builtin.copy:
+    src: "complex_variables/{{ item }}"
+    dest: "{{ terraform_project_dir }}/complex_vars/{{ item }}"
+  with_items:
+    - main.tf
+    - variables.tf
+
+# This task would test the various complex variable structures of the with the
+# terraform null_resource
+- name: test complex variables
+  community.general.terraform:
+    project_path: "{{ terraform_project_dir }}/complex_vars"
+    binary_path: "{{ terraform_binary_path }}"
+    force_init: yes
+    variables:
+      dictionaries:
+        name: kosala
+        age: 99
+      list_of_strings:
+        - kosala
+        - yyy
+        - xxx
+        - zzz
+      list_of_objects:
+        - name: kosala
+          age: 99
+        - name: yyy
+          age: 0.1
+        - name: zzz
+          age: 9.789
+        - name: lll
+          age: 1000
+      boolean: true
+      string_type: 'randomstring2"&$%@'
```

Something like this will be more interesting, I think:
```suggestion
      string_type: 'randomstring2"&$%@'
      string_type_2: 'random" string 2 '
```

---

The reason I had the string_type is to check whether we pass on the string without alteration from ansible to terraform without any deformation. I do not see the value of separating multiple characters to two strings, may be we can make it longer to represent all the characters :).

```
string_type: 'quickbrownfoxjumpoverthelazydog 1234567890 " &$#%@!~*^()[]{}'
```

Since we handle top level and within Dict / List strings differently may be we should perform the same check.

---

@kosalaat testing all characters brings little value. The purpose of that comment by Felix is that we should test if the module can handle strings containing:
- whitespaces
- the single quote character
- the double quote character
- combinations of these

---

Testing all characters does not bring much value, so refined the tests to the following.

1. Special shell charachters: `!@#$%&*(){}[]`
2. Spaces
3. Double quotes
4. Colon `:`

Tests are covered for top tier and second tier variables.